### PR TITLE
MONIT-34354: Proxy code changes to support integration with CSP for g…

### PR DESCRIPTION
**Pull request by Norayr25**
…etting an access token

### CODE REVIEW

File: proxy/pom.xml
Comment (laullon) [inline proxy/pom.xml:645]: why all this new dependencies ?

File: proxy/src/main/java/com/wavefront/agent/TenantInfo.java
Comment (laullon) [inline proxy/src/main/java/com/wavefront/agent/TenantInfo.java:126]: use RestEasy, like other parts of the Proxy

File: proxy/src/main/java/com/wavefront/agent/TenantInfo.java
Comment (laullon) [inline proxy/src/main/java/com/wavefront/agent/TenantInfo.java:77]: looks like duplicated code

File: proxy/src/main/java/com/wavefront/agent/TenantInfo.java
Comment (laullon) [inline proxy/src/main/java/com/wavefront/agent/TenantInfo.java:43]: if this will have only one item, why use a Caffeine cache, looks overkill.

File: proxy/src/main/java/com/wavefront/agent/TenantInfo.java
Comment (Norayr25) [reply proxy/src/main/java/com/wavefront/agent/TenantInfo.java:77]: > looks like duplicated code

No, there are two ways to build a tokencache: using getAccessTokenByAPIToken() or getAccessTokenByClientCredentials().

File: proxy/src/main/java/com/wavefront/agent/TenantInfo.java
Comment (Norayr25) [reply proxy/src/main/java/com/wavefront/agent/TenantInfo.java:43]: > if this will have only one item, why use a Caffeine cache, looks overkill.

We use Caffeine cache for regularly refreshing token if it comes from CSP based on token TTL.

**Comment by Norayr25** [comment]
> run `make build-jar` before committing, that will reformat the code, and fix the build action.

thanks

File: proxy/src/main/java/com/wavefront/agent/TenantInfo.java
Comment (Norayr25) [reply proxy/src/main/java/com/wavefront/agent/TenantInfo.java:77]: > yeah but you have 25 lines duplicated, and only change 1

ahh got it, i'll make changes

File: proxy/src/main/java/com/wavefront/agent/TenantInfo.java
Comment (ben-manes) [reply proxy/src/main/java/com/wavefront/agent/TenantInfo.java:43]: fyi, you might review `Suppliers.memoizeWithExpiration` which uses double checking locking with a TTL check. Their api has the TTL being static, but conceptually would not need to be in a custom variant.

File: proxy/src/main/java/com/wavefront/agent/TenantInfo.java
Comment (Norayr25) [reply proxy/src/main/java/com/wavefront/agent/TenantInfo.java]: Done

File: proxy/src/main/java/com/wavefront/agent/TenantInfo.java
Comment (Norayr25) [reply proxy/src/main/java/com/wavefront/agent/TenantInfo.java]: Done, thanks

File: proxy/pom.xml
Comment (laullon) [inline proxy/pom.xml:606]: Please, remove this, and the need to gets methods.
We already have too many dependencies, and this jus save you few very simples lines of code.

File: proxy/src/main/java/com/wavefront/agent/TenantInfo.java
Comment (laullon) [inline proxy/src/main/java/com/wavefront/agent/TenantInfo.java:139]: this will prevent the `loadAccessTokenByClientCredentials` to be scheduled.

File: proxy/src/main/java/com/wavefront/agent/ProxyConfig.java
Comment (laullon) [inline proxy/src/main/java/com/wavefront/agent/ProxyConfig.java:1483]: we should log, or crash, if 4 values are set.
Also, add a log info message indicating which method are we using.

File: proxy/src/main/java/com/wavefront/agent/TenantInfo.java
Comment (laullon) [inline proxy/src/main/java/com/wavefront/agent/TenantInfo.java:27]: this should be a config option, just in case.

File: proxy/src/main/java/com/wavefront/agent/TenantInfo.java
Comment (laullon) [inline proxy/src/main/java/com/wavefront/agent/TenantInfo.java:31]: this can be static, so we just have one thread, no one per tenant.

**Comment by laullon** [comment]
If the Token is expired (or not valid) the server will return 403 ?
https://github.com/wavefrontHQ/wavefront-proxy/blob/643c8f2a8ef93ab92de03abec24c091653d06465/proxy/src/main/java/com/wavefront/agent/data/AbstractDataSubmissionTask.java#L125

File: proxy/src/main/java/com/wavefront/agent/ProxyConfig.java
Comment (Norayr25) [reply proxy/src/main/java/com/wavefront/agent/ProxyConfig.java:1483]: > we should log, or crash, if 4 values are set. Also, add a log info message indicating which method are we using.

I'll add logs. If 4 values are set we decided to give priority to the clientId, clientSecret

File: proxy/src/main/java/com/wavefront/agent/TenantInfo.java
Comment (Norayr25) [reply proxy/src/main/java/com/wavefront/agent/TenantInfo.java:153]: Yes you are right, Will make change to make request a bit early

File: proxy/src/main/java/com/wavefront/agent/TenantInfo.java
Comment (Norayr25) [reply proxy/src/main/java/com/wavefront/agent/TenantInfo.java:139]: Yes, it is the happy way version. Another ticket will be created to have better error handling. Additionally, I'll add unit tests during that ticket before putting it up for review.

File: proxy/pom.xml
Comment (Norayr25) [reply proxy/pom.xml:606]: Done

**Comment by Norayr25** [comment]
> If the Token is expired (or not valid) the server will return 403 ?
> 
> https://github.com/wavefrontHQ/wavefront-proxy/blob/643c8f2a8ef93ab92de03abec24c091653d06465/proxy/src/main/java/com/wavefront/agent/data/AbstractDataSubmissionTask.java#L125

It will still be valid with the previous permissions if you mean that the CSP access token expired before you requested a new access token from the CSP. However, we will now request an access token a little bit earlier.

File: proxy/src/main/java/com/wavefront/agent/TenantInfo.java
Comment (laullon) [reply proxy/src/main/java/com/wavefront/agent/TenantInfo.java:138]: why another ticket ? is this PR just for testing?

File: proxy/src/main/java/com/wavefront/agent/ProxyConfigDef.java
Comment (laullon) [inline proxy/src/main/java/com/wavefront/agent/ProxyConfigDef.java:1519]: this should be a config option.

File: proxy/src/main/java/com/wavefront/agent/TenantInfo.java
Comment (laullon) [inline proxy/src/main/java/com/wavefront/agent/TenantInfo.java:109]: duplicated code

File: proxy/src/main/java/com/wavefront/agent/TenantInfo.java
Comment (laullon) [inline proxy/src/main/java/com/wavefront/agent/TenantInfo.java:91]: why retry only 3 times ?
I think this should be implemented on a simpler way..
If we get token schedule in `getExpiresIn` if not schedule in 1 min.

File: proxy/src/main/java/com/wavefront/agent/TenantInfo.java
Comment (laullon) [inline proxy/src/main/java/com/wavefront/agent/TenantInfo.java:185]: duplicated

Comment (Norayr25) [reply proxy/src/main/java/com/wavefront/agent/TenantInfo.java]: done

File: proxy/src/main/java/com/wavefront/agent/TenantInfo.java
Comment (Norayr25) [reply proxy/src/main/java/com/wavefront/agent/TenantInfo.java]: done

File: proxy/src/main/java/com/wavefront/agent/TenantInfo.java
Comment (Norayr25) [reply proxy/src/main/java/com/wavefront/agent/TenantInfo.java]: done

File: proxy/src/main/java/com/wavefront/agent/ProxyConfigDef.java
Comment (akodali18) [inline proxy/src/main/java/com/wavefront/agent/ProxyConfigDef.java:1522]: @laullon i just happened to check this PR out as our team was waiting on the proxy release to support csp auth in the opApps k8s operator. Quick question, IIUC, this would translate to a user config option in the wavefront.conf file. In that case, to maintain consistency with other proxy config options,  this can be named in camel case. Ex: "--cspApiToken" . 
Same comment for other newly added csp specific config parameters.

Comment (laullon) [reply proxy/src/main/java/com/wavefront/agent/ProxyConfigDef.java:1522]: sure. no problem

File: proxy/src/main/java/com/wavefront/agent/ProxyConfigDef.java
Comment (laullon) [reply proxy/src/main/java/com/wavefront/agent/ProxyConfigDef.java:1519]: ^

**Comment by ernst-riemer** [comment]
@Norayr25 @laullon We're looking at what changes we need to do on our side for our tas2to product. Do we need to have orgId as part of the proxy configuration? I think orgId is in both user and server-to-server csp jwt tokens. Open to hop on a zoom to chat if needed.

**Comment by Norayr25** [comment]
> @Norayr25 @laullon We're looking at what changes we need to do on our side for our tas2to product. Do we need to have orgId as part of the proxy configuration? I think orgId is in both user and server-to-server csp jwt tokens. Open to hop on a zoom to chat if needed.

@ernst-riemer The orgId is required when we use OAuth app(appId, appSecret) for authentication

File: pkg/etc/wavefront/wavefront-proxy/wavefront.conf.default
Comment (laullon) [inline pkg/etc/wavefront/wavefront-proxy/wavefront.conf.default:41]: this doesn't match the default value, and commented

File: pkg/etc/wavefront/wavefront-proxy/wavefront.conf.default
Comment (laullon) [inline pkg/etc/wavefront/wavefront-proxy/wavefront.conf.default:25]: this should be commented

File: pkg/etc/wavefront/wavefront-proxy/wavefront.conf.default
Comment (laullon) [inline pkg/etc/wavefront/wavefront-proxy/wavefront.conf.default:30]: same

File: pkg/etc/wavefront/wavefront-proxy/wavefront.conf.default
Comment (laullon) [inline pkg/etc/wavefront/wavefront-proxy/wavefront.conf.default:35]: same

**Review by laullon — CHANGES_REQUESTED** [review]

File: pkg/etc/wavefront/wavefront-proxy/wavefront.conf.default
Comment (Norayr25) [reply pkg/etc/wavefront/wavefront-proxy/wavefront.conf.default]: > this doesn't match the default value, and commented

fixed

**Review by laullon — APPROVED** [review]
LGTM

File: docker/Dockerfile
Comment (joannak-vmware) [inline docker/Dockerfile:6]: nit: capitalize Wavefront

File: docker/Dockerfile-rhel
Comment (joannak-vmware) [inline docker/Dockerfile-rhel:19]: nit: same here

File: proxy/src/main/java/com/wavefront/agent/api/APIContainer.java
Comment (joannak-vmware) [inline proxy/src/main/java/com/wavefront/agent/api/APIContainer.java:85]: Has this TODO been addressed?

**Comment by joannak-vmware** [comment]
Nit comments, else LGTM. Will `Approve` PR after QA Validated.

File: docker/Dockerfile
Comment (Norayr25) [reply docker/Dockerfile:6]: > nit: capitalize Wavefront

changed

File: docker/run.sh
Comment (Norayr25) [reply docker/run.sh:18]: > Can we clarify message, invalid combination of token parameters?

changed

File: proxy/src/main/java/com/wavefront/agent/api/APIContainer.java
Comment (Norayr25) [reply proxy/src/main/java/com/wavefront/agent/api/APIContainer.java]: > Has this TODO been addressed?

German did refactoring

File: pkg/etc/wavefront/wavefront-proxy/wavefront.conf.default
Comment (Norayr25) [reply pkg/etc/wavefront/wavefront-proxy/wavefront.conf.default:21]: done

**Comment by ernst-riemer** [comment]
We're adding some changes to make orgId an optional parameter for the proxy config. This was discussed with Garnik Gevorgyan. We also added some tests & fixed a couple of typos. 

When an orgId is not specified in the wavefront config, the default csp behavior will return a JWT token with the orgId that "owns" the oauth app, and that oauth app "owns" the client credentials.

These changes still allow for setting the orgId in the wavefront proxy config file, but is not required by default.

**Comment by jbooherl** [comment]
I was able to merge in Ernst's PR to this PR. @markdouglasswork can address any changes that we need to make if there are comments.

File: docker/run.sh
Comment (laullon) [inline docker/run.sh:10]: CSP_ORG_ID is now optional

Comment (ernst-riemer) [reply docker/run.sh]: updated

Comment (jbooherl) [reply docker/run.sh]: Good catch! Updated that, the readme and comments

**Review by markdouglasswork — APPROVED** [review]
CPO Team Troll switched approval of this to me since @ernst-riemer and @jbooherl are out. This looks good to me now that the org ID is optional.

**Review by joannak-vmware — APPROVED** [review]
LGTM

**Pull request merged**
